edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+    sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)
+
+    # TCP_KEEPIDLE: the time (in seconds) the connection needs to remain idle
+    # before TCP starts sending keepalive probes. This is socket.TCP_KEEPIDLE
+    # on Linux, and socket.TCP_KEEPALIVE on macOS from Python 3.10.
+    if hasattr(socket, 'TCP_KEEPIDLE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPIDLE, TCP_KEEPIDLE)
+    if hasattr(socket, 'TCP_KEEPALIVE'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPALIVE, TCP_KEEPIDLE)
+
+    # TCP_KEEPINTVL: The time (in seconds) between individual keepalive probes.
+    if hasattr(socket, 'TCP_KEEPINTVL'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPINTVL, TCP_KEEPINTVL)
+
+    # TCP_KEEPCNT: The maximum number of keepalive probes TCP should send
+    # before dropping the connection.
+    if hasattr(socket, 'TCP_KEEPCNT'):
+        sock.setsockopt(socket.IPPROTO_TCP,
+                        socket.TCP_KEEPCNT, TCP_KEEPCNT)
+
+
+P = TypeVar('P', bound=asyncio.Protocol)
+
+
+async def create_postgres_connection(
+    dsn: str | ConnectionParams,
+    protocol_factory: Callable[[], P],
+    *,
+    source_description: Optional[str] = None,
+    state_change_callback: Optional[StateChangeCallback] = None
+) -> Tuple[PGRawConn, P]:
+    """
+    Open a PostgreSQL connection to the address specified by the DSN or
+    ConnectionParams, creating the user protocol from the protocol_factory.
+
+    This method establishes the connection asynchronously. When successful, it
+    returns a (PGRawConn, protocol) pair.
+    """
+    if isinstance(dsn, str):
+        dsn = ConnectionParams(dsn=dsn)
+    connect_timeout = dsn.connect_timeout
+    try:
+        state = PyConnectionState(
+            dsn._params,
+            "postgres",
+            str(get_pg_home_directory())
+        )
+    except Exception as e:
+        raise ValueError(e)
+    ready_future: asyncio.Future = asyncio.Future()
+    pg_state = PGState(
+        parameters={},
+        cancellation_key=None,
+        auth=None,
+        server_error=None,
+        ssl=False
+    )
+
+    user_protocol = protocol_factory()
+    protocol, host, port, initial_transport = await _create_connection(
+        lambda: PGConnectionProtocol(
+            state,
+            user_protocol,
+            ready_future,
+            pg_state
+        ),
+        connect_timeout,
+        state.config.host_candidates
+    )
+
+    upgraded_transport = None
+    try:
+        update = RustTransportUpdate(
+            state,
+            initial_transport,
+            state_change_callback,
+            pg_state,
+            host if protocol != "unix" else None,
+            ready_future
+        )
+        state.update = update
+        state.drive_initial()
+
+        await ready_future
+        upgraded_transport = update.raw_transport
+        conn = PGRawConn(
```

Comment:
@fantix I wonder if it would make sense to just return a raw transport here rather than having `PGRawConn`. The only question I have is how we might be able to expose the DB state (connection params, etc).

Reply:
I think in `PGConnection.connection_made()`, we will have the `PGRawConn` and we can just setup `PGConnection` properties with values from `PGRawConn` there.